File: benchmarks/cross-language/cpp/skynet.cpp
/**
 * C++ Skynet Benchmark
 * Based on https://github.com/atemerev/skynet
 * Uses std::thread for top THREAD_DEPTH levels, sequential below.
 * Spawning 1M OS threads is not feasible; limits concurrent threads to ~1000.
 *
 * Compile with: g++ -O3 -std=c++17 -march=native skynet.cpp -o skynet -pthread
 */

#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <chrono>
#include <cstdlib>
#include <cstring>
#include <array>

// Below THREAD_DEPTH, compute the sub-tree sum sequentially on the current thread.
static const int THREAD_DEPTH = 3;

static long long skynet_seq(long long offset, long long size) {
    if (size == 1) return offset;
    long long child_size = size / 10;
    long long sum = 0;
    for (int i = 0; i < 10; i++) {
        sum += skynet_seq(offset + (long long)i * child_size, child_size);
    }
    return sum;
}

// Each node sends its result to a mutex-protected slot, then signals its parent.
struct ResultSlot {
    long long value = 0;
    std::mutex mtx;
    std::condition_variable cv;
    bool ready = false;
};

static void skynet(ResultSlot* out, long long offset, long long size, int depth) {
    if (size == 1 || depth >= THREAD_DEPTH) {
        std::lock_guard<std::mutex> lock(out->mtx);
        out->value = skynet_seq(offset, size);
        out->ready = true;
        out->cv.notify_one();
        return;
    }
    long long child_size = size / 10;
    std::array<ResultSlot, 10> slots;
    std::array<std::thread, 10> threads;

    for (int i = 0; i < 10; i++) {
        long long child_offset = offset + (long long)i * child_size;
        ResultSlot* slot = &slots[i];
        int next_depth = depth + 1;
        threads[i] = std::thread([slot, child_offset, child_size, next_depth]() {
            skynet(slot, child_offset, child_size, next_depth);
        });
    }

    long long sum = 0;
    for (int i = 0; i < 10; i++) {
        threads[i].join();
        sum += slots[i].value;
    }

    std::lock_guard<std::mutex> lock(out->mtx);
    out->value = sum;
    out->ready = true;
    out->cv.notify_one();
}

static long long get_leaves() {
    const char* env = getenv("SKYNET_LEAVES");
    if (env) return atoll(env);
    env = getenv("BENCHMARK_MESSAGES");
    if (env) return atoll(env);
    return 1000000LL;
}

int main() {
    long long num_leaves = get_leaves();

    // Total actors = sum of nodes at each level
    long long total_actors = 0;
    for (long long n = num_leaves; n >= 1; n /= 10) {
        total_actors += n;
    }

    std::cout << "=== C++ Skynet Benchmark ===" << std::endl;
    std::cout << "Leaves: " << num_leaves << " (std::thread, top " << THREAD_DEPTH << " levels parallel)" << std::endl;
    std::cout << std::endl;

    ResultSlot root;
    auto start = std::chrono::high_resolution_clock::now();
    std::thread t([&root, num_leaves]() { skynet(&root, 0, num_leaves, 0); });
    t.join();
    auto end = std::chrono::high_resolution_clock::now();

    long long elapsed_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    long long elapsed_us = elapsed_ns / 1000;

    std::cout << "Sum: " << root.value << std::endl;
    if (elapsed_us > 0) {
        long long ns_per_msg = elapsed_ns / total_actors;
        long long throughput_m = total_actors / elapsed_us;
        long long leftover = total_actors - (throughput_m * elapsed_us);
        long long throughput_frac = (leftover * 100) / elapsed_us;
        std::cout << "ns/msg:         " << ns_per_msg << std::endl;
        std::cout << "Throughput:     " << throughput_m << ".";
        if (throughput_frac < 10) std::cout << "0";
        std::cout << throughput_frac << " M msg/sec" << std::endl;
    }
    return 0;
}

File: benchmarks/cross-language/rust/src/skynet.rs
// Rust Skynet Benchmark
// Based on https://github.com/atemerev/skynet
// Uses std::thread for top levels (THREAD_DEPTH levels), sequential below.
// Spawning 1M OS threads is not feasible; this limits concurrent threads to ~1000.

use std::sync::mpsc::{channel, Sender};
use std::thread;
use std::time::Instant;

fn get_leaves() -> i64 {
    if let Ok(s) = std::env::var("SKYNET_LEAVES") {
        if let Ok(n) = s.parse() {
            return n;
        }
    }
    if let Ok(s) = std::env::var("BENCHMARK_MESSAGES") {
        if let Ok(n) = s.parse() {
            return n;
        }
    }
    1_000_000
}

// Below THREAD_DEPTH, compute the sub-tree sum sequentially on the current thread.
const THREAD_DEPTH: usize = 3;

fn skynet_seq(offset: i64, size: i64) -> i64 {
    if size == 1 {
        return offset;
    }
    let child_size = size / 10;
    let mut sum = 0i64;
    for i in 0..10i64 {
        sum += skynet_seq(offset + i * child_size, child_size);
    }
    sum
}

fn skynet(tx: Sender<i64>, offset: i64, size: i64, depth: usize) {
    if size == 1 || depth >= THREAD_DEPTH {
        tx.send(skynet_seq(offset, size)).unwrap();
        return;
    }
    let child_size = size / 10;
    let (child_tx, child_rx) = channel::<i64>();
    for i in 0..10i64 {
        let child_offset = offset + i * child_size;
        let child_tx = child_tx.clone();
        thread::spawn(move || {
            skynet(child_tx, child_offset, child_size, depth + 1);
        });
    }
    let mut sum = 0i64;
    for _ in 0..10 {
        sum += child_rx.recv().unwrap();
    }
    tx.send(sum).unwrap();
}

fn main() {
    let num_leaves = get_leaves();

    // Total actors = sum of nodes at each level
    let mut total_actors = 0i64;
    let mut n = num_leaves;
    while n >= 1 {
        total_actors += n;
        n /= 10;
    }

    println!("=== Rust Skynet Benchmark ===");
    println!("Leaves: {} (std::thread, top {} levels parallel)\n", num_leaves, THREAD_DEPTH);

    let (tx, rx) = channel::<i64>();
    let start = Instant::now();
    thread::spawn(move || skynet(tx, 0, num_leaves, 0));
    let sum = rx.recv().unwrap();
    let elapsed = start.elapsed();

    let elapsed_ns = elapsed.as_nanos() as i64;
    let elapsed_us = elapsed_ns / 1000;

    println!("Sum: {}", sum);
    if elapsed_us > 0 {
        let ns_per_msg = elapsed_ns / total_actors;
        let throughput_m = total_actors / elapsed_us;
        let leftover = total_actors - (throughput_m * elapsed_us);
        let throughput_frac = (leftover * 100) / elapsed_us;
        print!("ns/msg:         {}\n", ns_per_msg);
        print!("Throughput:     {}.", throughput_m);
        if throughput_frac < 10 {
            print!("0");
        }
        println!("{} M msg/sec", throughput_frac);
    }
}
